File: receiver.c
#include "buffer.c"

enum STATE { CLOSED, LISTEN, SYN_RECD, ESTABLISHED, CLOSE_WAIT, LAST_ACK };

int seq_num = 154;
int ack_num = 0;
enum STATE state = CLOSED;

int data_transferred = 0;
int data_segments = 0;
int duplicate_segments = 0;

int port;
const char* file;

const char* log_file = "Receiver_log.txt";

int main(int argc, char* argv[]) {
  if (argc != 3) {
    printf("Usage: receiver.py receiver_port file.txt");
    return -1;
  }

  port = atoi(argv[1]);
  file = argv[2];

  FILE* fp = fopen(file, "w");
  fclose(fp);

  while (1) {
    if (state == CLOSED) {
      stp_init();
      stp_bind(port);
      state = LISTEN;
    }

    if (state == LISTEN) {
      struct stp_vo* pkt = stp_recv();

      if (pkt->packet.SYN) {
        ack_num = pkt->packet.seq_num + 1;
        free(pkt);

        stp_send_empty_pkt(seq_num, ack_num, 1, 1, 0);
        seq_num += 1;
        state = SYN_RECD;
      }
    }

    if (state == SYN_RECD) {
      struct stp_vo* pkt = stp_recv();

      if (pkt->packet.ACK && pkt->packet.ack_num == seq_num) {
        free(pkt);
        state = ESTABLISHED;
      }
    }

    if (state == ESTABLISHED) {
      struct stp_vo* pkt = stp_recv();

      if (!pkt->packet.FIN) {
        if (pkt->packet.seq_num < ack_num || buffer_in(pkt->packet.seq_num) != -1) {
          free(pkt);
          duplicate_segments += 1;
          continue;
        }
        data_segments += 1;
        buffer_add(pkt);
        int pos;
        while ((pos = buffer_in(ack_num)) != -1) {
          struct stp_vo* pkt = buffer_pop(pos);
          ack_num += pkt->data_size;
          data_transferred += pkt->data_size;
          FILE* fp = fopen(file, "a");
          fwrite(pkt->packet.data, 1, pkt->data_size, fp);
          fclose(fp);
          free(pkt);
        }

        stp_send_empty_pkt(seq_num, ack_num, 0, 1, 0);
      } else {
        free(pkt);
        state = CLOSE_WAIT;
      }
    }

    if (state == CLOSE_WAIT) {
      ack_num += 1;
      stp_send_empty_pkt(seq_num, ack_num, 0, 1, 1);
      seq_num += 1;
      state = LAST_ACK;
    }

    if (state == LAST_ACK) {
      struct stp_vo* pkt = stp_recv();

      if (pkt->packet.ACK && pkt->packet.ack_num == seq_num) {
        stp_close();
        state = CLOSED;
        break;
      }
    }
  }

    sprintf(log_buffer, "\n");
    stp_log();
    sprintf(log_buffer, "Amount of Data Received: %d", data_transferred);
    stp_log();
    sprintf(log_buffer, "Number of Data Segments Received: %d", data_segments);
    stp_log();
    sprintf(log_buffer, "Number of duplicate segments received: %d", duplicate_segments);
    stp_log();
}

File: v1_py/receiver.py
#!/usr/bin/env python

########################################
#  Simple Transport Protocol Receiver  #
#        sjysd(Tentative)  2019/05/07  #
########################################

import sys

from STP import *

if len(sys.argv) != 3:
    print("Usage: receiver.py receiver_port file.txt")
    sys.exit(1)
port, file = sys.argv[1:]
port = int(port)

addr = None

seq_num = 100
ack_num = 0
state = "CLOSED"

data_transferred = 0
data_segments = 0
duplicate_segments = 0

receiver = STPServer("Receiver_log.txt")
buffer = {}

open(file, "w").close()

while True:
    if state == "CLOSED":
        receiver.bind(port)
        state = "LISTEN"

    if state == "LISTEN":
        pkt, addr = receiver.recv_with_udp_header()

        if pkt.syn:
            ack_num = pkt.seq_num + 1
            receiver.send(STPPacket(seq_num, ack_num, syn=True, ack=True), addr)
            seq_num += 1
            state = "SYN_RECD"

    if state == "SYN_RECD":
        pkt = receiver.recv()

        if pkt.ack and pkt.ack_num == seq_num:
            state = "ESTABLISHED"

    if state == "ESTABLISHED":
        pkt = receiver.recv()

        if not pkt.fin:
            if pkt.seq_num < ack_num or pkt.seq_num in buffer:
                duplicate_segments += 1
                continue
            data_segments += 1
            buffer[pkt.seq_num] = pkt
            while ack_num in buffer:
                pkt = buffer.pop(ack_num)
                ack_num += len(pkt.data)
                data_transferred += len(pkt.data)
                f = open(file, "a+b")
                f.write(pkt.data)
                f.close()
            receiver.send(STPPacket(seq_num, ack_num, ack=True), addr)

        else:
            state = "CLOSE_WAIT"

    if state == "CLOSE_WAIT":
        ack_num += 1
        receiver.send(STPPacket(seq_num, ack_num, ack=True, fin=True), addr)
        seq_num += 1
        state = "LAST_ACK"

    if state == "LAST_ACK":
        pkt = receiver.recv()

        if pkt.ack and pkt.ack_num == seq_num:
            receiver.close()
            state = "CLOSED"
            break

receiver.log("")
receiver.log("Amount of Data Received " + str(data_transferred))
receiver.log("Number of Data Segments Received: " + str(data_segments))
receiver.log("Number of duplicate segments received: " + str(duplicate_segments))
